Document mobile game automation (#3527)
saucelabs/sauce-docs: docs/mobile-apps/automated-testing/alttester.md

---
id: alttester
title: AltTester for Mobile Game Automation
sidebar_label: Overview
hide_table_of_contents: true
---

import useBaseUrl from '@docusaurus/useBaseUrl';

> [AltTester](https://alttester.com/) is an open-source UI test automation tool for **game engines**. It instruments your build with a small SDK that exposes the in-game scene graph to C# tests over WebSocket, so tests can find GameObjects (Unity) or actors and widgets (Unreal) by name and drive real input on them. Sauce Labs supports AltTester for both Unity and Unreal Engine on real iOS and Android devices.

Reviewer: ```suggestion
[AltTester](https://alttester.com/) is an open-source test automation framework for Unity and Unreal Engine games. It instruments your build so that automated tests can inspect the live scene graph, simulate user input, and assert on game state in real time. Tests communicate with the instrumented build over WebSocket and can be written in C#, Java, Python or Robot Framework. Sauce Labs supports AltTester for both Unity and Unreal Engine on real iOS and Android devices.
```

## How It Works with Sauce Labs

```text
+-------------------+        +-------------------+        +----------------------+
|  C# test (NUnit)  |  WS    |  AltTester        |  WS    |  Instrumented game   |
|  AltTester-Driver | <----> |  Desktop or       | <----> |  on Sauce real       |
|  Appium.WebDriver |  HTTP  |  Sauce Connect    |        |  device (Appium)     |

> +-------------------+        +-------------------+        +----------------------+

Reviewer: ```suggestion
+-------------------+        +-------------------+        +----------------------+
|                      |  WS / WSS    |  AltTester        |  WS / WSS    |  Instrumented game   |
|  AltTester-Driver | <----> |  Desktop or       | <----> |  on Sauce real       |
|  Appium.WebDriver |  HTTP |  Sauce Connect    |        |  device (Appium)     |
+-------------------+        +-------------------+        +----------------------+
```

> 
> ```text
> +-------------------+              +-------------------+              +----------------------+
> |  C# test (NUnit)  |  WS / WSS    |  AltTester®       |  WS / WSS    |  Instrumented game   |
> |  AltTester-Driver | <----------> |  Desktop or       | <----------> |  on Sauce real       |
> |  Appium.WebDriver |    HTTP      |  Sauce Connect    |              |  device (Appium)     |
> +-------------------+              +-------------------+              +----------------------+
> ```

Reviewer: ```suggestion
```text
+-------------------+              +-------------------+              +----------------------+
|  AltTester®          |  WS / WSS    |  AltTester®       |  WS / WSS    |  Instrumented game   |
|  Driver and          | <----------> |  Desktop or       | <----------> |  on Sauce real       |
|  Appium.WebDriver |    HTTP      |  Sauce Connect    |              |  device (Appium)     |
+-------------------+              +-------------------+              +----------------------+
```
```

1. You instrument your Unity or Unreal build with the AltTester® SDK and produce a `.apk` or `.ipa`.
2. You upload that build to **Sauce App Storage**.
3. Your C# test starts an **Appium** session on a Sauce real device, which installs and launches the game.

> 4. Once the game is running, your test connects to the **AltTester® C# driver**, which talks to the in-game SDK over WebSocket (tunneled through Sauce Connect Proxy or a public VM).

Reviewer: ```suggestion
4. Once the game is running, your test connects to the **AltTester® Driver**, which talks to the in-game SDK over WebSocket (tunneled through Sauce Connect Proxy or a public VM).
```
